File: backend/app/schemas/employee.py
from fastapi.middleware.cors import CORSMiddleware
from decimal import Decimal
from datetime import datetime, date
from typing import Optional, List, Any
from pydantic import BaseModel, ConfigDict, Field, model_validator, AliasPath
from pydantic.alias_generators import to_camel
from pydantic import AliasChoices

# Helper to accept both snake_case and camelCase
def _camel_alias(field_name: str) -> str:
    return to_camel(field_name)

class EmployeeBase(BaseModel):
    full_name: str
    display_name: Optional[str] = None
    phone_primary: str
    phone_secondary: Optional[str] = None
    profile_image_url: Optional[str] = None
    national_id: Optional[str] = None
    birth_date: Optional[datetime] = None
    governorate: Optional[str] = None
    city: Optional[str] = None
    detailed_address: Optional[str] = None
    personal_notes: Optional[str] = None
    bio_ar: Optional[str] = None
    bio_en: Optional[str] = None

    job_title: str = "barber"
    department: Optional[str] = None
    employment_type: str = "full_time"
    hire_date: Optional[datetime] = None
    status: str = "active"
    work_days_json: Optional[List[int]] = None
    work_hours_json: Optional[Any] = None
    
    show_in_pos: bool = True
    show_in_booking: bool = True
    display_order: int = 0
    is_active: bool = True

    base_salary: Decimal = Decimal("0.00")
    commission_rate: Decimal = Decimal("0.00")
    fixed_bonus: Decimal = Decimal("0.00")
    bonus_min_attendance_percent: Decimal = Decimal("0.00")
    enable_attendance_auto_deduction: bool = True
    discipline_bonus: Decimal = Decimal("0.00")
    default_deductions: Decimal = Decimal("0.00")
    payment_method: Optional[str] = None
    wallet_number: Optional[str] = None
    bank_account: Optional[str] = None

    assistant_of_barber_id: Optional[int] = None
    assistant_tasks_json: Optional[List[str]] = None
    receives_commission: bool = False
    assistant_commission_rate: Decimal = Decimal("0.00")

    has_login_account: bool = False
    user_id: Optional[int] = None

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
        alias_generator=to_camel,
        extra="ignore",
    )

class EmployeeCreate(EmployeeBase):
    # Allow both camelCase and snake_case on creation
    # Extended fields for optional login creation (handled in endpoint, not stored in employees table directly beyond has_login_account/user_id)
    username: Optional[str] = None
    password: Optional[str] = None
    role: Optional[str] = None
    service_ids: Optional[List[int]] = Field(default=None, alias="serviceIds")  # allow frontend serviceIds
    # also accept serviceIds snake
    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
        alias_generator=to_camel,
        extra="ignore",
    )

class EmployeeUpdate(BaseModel):
    full_name: Optional[str] = None
    display_name: Optional[str] = None
    phone_primary: Optional[str] = None
    phone_secondary: Optional[str] = None
    profile_image_url: Optional[str] = None
    national_id: Optional[str] = None
    birth_date: Optional[datetime] = None
    governorate: Optional[str] = None
    city: Optional[str] = None
    detailed_address: Optional[str] = None
    personal_notes: Optional[str] = None
    bio_ar: Optional[str] = None
    bio_en: Optional[str] = None

    job_title: Optional[str] = None
    department: Optional[str] = None
    employment_type: Optional[str] = None
    hire_date: Optional[datetime] = None
    status: Optional[str] = None
    work_days_json: Optional[List[int]] = None
    work_hours_json: Optional[Any] = None
    
    show_in_pos: Optional[bool] = None
    show_in_booking: Optional[bool] = None
    display_order: Optional[int] = None
    is_active: Optional[bool] = None

    base_salary: Optional[Decimal] = None
    commission_rate: Optional[Decimal] = None
    fixed_bonus: Optional[Decimal] = None
    bonus_min_attendance_percent: Optional[Decimal] = None
    enable_attendance_auto_deduction: Optional[bool] = None
    discipline_bonus: Optional[Decimal] = None
    default_deductions: Optional[Decimal] = None
    payment_method: Optional[str] = None
    wallet_number: Optional[str] = None
    bank_account: Optional[str] = None

    assistant_of_barber_id: Optional[int] = None
    assistant_tasks_json: Optional[List[str]] = None
    receives_commission: Optional[bool] = None
    assistant_commission_rate: Optional[Decimal] = None

    has_login_account: Optional[bool] = None
    user_id: Optional[int] = None
    username: Optional[str] = None
    password: Optional[str] = None
    role: Optional[str] = None
    service_ids: Optional[List[int]] = Field(default=None, alias="serviceIds")

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
        alias_generator=to_camel,
        extra="ignore",
    )

from app.schemas.service import ServiceRead

class EmployeeRead(EmployeeBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    # Computed or linked fields
    assistant_of_name: Optional[str] = None
    services: List[ServiceRead] = []

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
        alias_generator=to_camel,
        extra="ignore",
    )

class EmployeeListItem(BaseModel):
    id: int
    full_name: str
    display_name: Optional[str] = None
    job_title: str
    phone_primary: str
    status: str
    is_active: bool
    base_salary: Decimal
    commission_rate: Decimal
    assistant_of_name: Optional[str] = None
    profile_image_url: Optional[str] = None
    service_ids: List[int] = []

    model_config = ConfigDict(
        from_attributes=True
    )

File: frontend/src/types/employee.ts
/**
 * Canonical employee record — merged from HR, Payroll & Attendance pages.
 * All fields optional: the backend returns mixed snake_case / camelCase shapes.
 */
export interface EmployeeRecord {
  id?: number | string;
  employee_id?: number | string;
  employeeId?: number | string;
  fullName?: string;
  full_name?: string;
  displayName?: string;
  display_name?: string;
  name?: string;
  status?: string;
  isActive?: boolean;
  is_active?: boolean;
  jobTitle?: string;
  job_title?: string;
  role?: string;
  phonePrimary?: string;
  phone_primary?: string;
  phoneSecondary?: string;
  phone_secondary?: string;
  profileImageUrl?: string;
  profile_image_url?: string;
  bioAr?: string;
  bio_ar?: string;
  bioEn?: string;
  bio_en?: string;
  nationalId?: string;
  national_id?: string;
  birthDate?: string;
  birth_date?: string;
  governorate?: string;
  city?: string;
  detailedAddress?: string;
  detailed_address?: string;
  personalNotes?: string;
  personal_notes?: string;
  department?: string;
  employmentType?: string;
  employment_type?: string;
  hireDate?: string;
  hire_date?: string;
  created_at?: string;
  showInPos?: boolean;
  show_in_pos?: boolean;
  showInBooking?: boolean;
  show_in_booking?: boolean;
  displayOrder?: number | string;
  display_order?: number | string;
  baseSalary?: number | string;
  base_salary?: number | string;
  commissionRate?: number | string;
  commission_rate?: number | string;
  fixedBonus?: number | string;
  fixed_bonus?: number | string;
  defaultDeductions?: number | string;
  default_ded_uctions?: number | string;
  paymentMethod?: string;
  payment_method?: string;
  walletNumber?: string;
  wallet_number?: string;
  bankAccount?: string;
  bank_account?: string;
  assistantOfBarberId?: number | string;
  assistant_of_barber_id?: number | string;
  assistantTasksJson?: unknown[];
  assistant_tasks_json?: unknown[];
  receivesCommission?: boolean;
  receives_commission?: boolean;
  assistantCommissionRate?: number | string;
  assistant_commission_rate?: number | string;
  hasLoginAccount?: boolean;
  has_login_account?: boolean;
  username?: string;
  password?: string;
  user_id?: number | string;
  services?: { id: number | string }[];
  serviceIds?: (number | string)[];
  [key: string]: unknown;
}

/**
 * Managed user row (moved from pages/owner/PermissionsManagement.tsx).
 */
export interface ManagedUser {
  id: string | number;
  fullName?: string;
  username?: string;
  role?: string;
  status?: string;
  isActive?: boolean;
  phone?: string;
  specialty?: string;
}

export interface DocumentRecord {
    id?: number | string;
    employee_id?: number | string;
    employeeId?: number | string;
    employeeName?: string;
    title?: string;
    file_type?: string;
    document_type?: string;
    documentType?: string;
  file_url?: string;
  fileUrl?: string;
  issue_date?: string;
  issueDate?: string;
  expiry_date?: string;
  expiryDate?: string;
  status?: string;
  notes?: string;
  created_at?: string;
  [key: string]: unknown;
}
